### ISSUE TITLE
BestOf2NearestRangeMatcher not working in Python or stitching_detailed.cpp

### ISSUE THREAD

**Issue by chinery** [opencv/opencv#22315]
##### System information (version)
- OpenCV => 4.6.0
- Operating System / Platform => MacOS
- Compiler => clang/Python

##### Detailed description

When calling `BestOf2NearestRangeMatcher` from Python with a `range_width` of 1, I still get matches between images that are not adjacent going by their index with non-zero confidence. In fact, changing the `range_width` value has no effect.

After downloading the source and running `stitching_detailed.cpp` I found the same issue.

##### Steps to reproduce

in Python:
```python
#... get features from a feature detector
range = 1
matcher = cv2.detail_BestOf2NearestRangeMatcher(range)
pairwise_matches = matcher.apply2(features)
print([(match.src_img_idx, match.dst_img_idx, match.confidence) for match in pairwise_matches])
```

for C++, run [`stitching_detailed.cpp`](https://github.com/opencv/opencv/blob/4.x/samples/cpp/stitching_detailed.cpp) with `--rangewidth` set to 1, and investigate the matches, e.g. add this to line 515

```c++
for (auto match : pairwise_matches) {
    cout << match.src_img_idx << "," << match.dst_img_idx << ": " << match.confidence << endl;
}
```

*Expected*: match confidences should mostly be zero except when `abs(src_idx - dst_idx) == 1`
*Actual*: match confidences are unchanged from when `range_width = -1`

##### *Potential fix!*

I think the issue in the C++ code is simply that the function is not marked as `virtual`, but is being called from a `Ptr<FeaturesMatcher>` in `stitching_detailed.cpp`. 

If you make the following change in [`matchers.hpp:140`](https://github.com/opencv/opencv/blob/38788a31619b742f3af86db4e0e5e9fcf8343de9/modules/stitching/include/opencv2/stitching/detail/matchers.hpp#L140):

`CV_WRAP_AS(apply2) void operator ()(const ...`

becomes

`CV_WRAP_AS(apply2) virtual void operator ()(const ...`

then `stitching_detailed.cpp` works as expected. 

(Probably a good idea to change [`matchers.hpp:127`](https://github.com/opencv/opencv/blob/38788a31619b742f3af86db4e0e5e9fcf8343de9/modules/stitching/include/opencv2/stitching/detail/matchers.hpp#L127) similarly)

I'd be happy to open a Pull Request for this fix. However, the stitching test suite does not run nicely on my M1 Mac and I haven't put much effort into fixing that yet.

I'm not sure why this polymorphism issue would affect Python where you call `cv2.detail_BestOf2NearestRangeMatcher` directly. Perhaps someone with more knowledge of how the Python to OpenCV layer works could help there.

##### Issue submission checklist

 - [X] I report the issue, it's not a question
   
 - [X] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found any solution
   
 - [X] I updated to the latest OpenCV version and the issue is still there
   
 - [X] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by chinery** [comment]
Just been reading the code a bit more, and realised that other parts of the stitching pipeline don't make the `operator ()` function virtual either. 

Instead, in e.g. `motion_estimators.hpp`, in the base class `Estimator`, `operator ()` calls `estimate`, and `estimate` is virtual. So there things work as expected.

I'm just not familiar enough with the code (or C++ to be honest!) to understand what should happen inside `matchers.hpp`. There are two versions of `operator ()`, the first one calls `match` which is labelled as `virtual`, however `BestOf2NearestRangeMatcher` does not override `match`. The second version of `operator ()` is implemented directly by `FeaturesMatcher`, not marked virtual, and _this_ is overridden by `BestOf2NearestRangeMatcher`.

**Comment by chinery** [comment]
As a final note on this from uneducated me... in terms of the Python bindings not working, I think the alternative to making the operator virtual might be to add a new `CV_WRAP_AS(apply2)` (or similar) to the BestOf2NearestRangeMatcher subclass version of that operator. The lack of that annotation is my best guess as to why calling `.apply2` doesn't work even when instantiating the subclass directly in Python.

**Comment by nndei** [comment]
Hello, has this been fixed in 4.7 update? Like mentioned [here](https://github.com/OpenStitching/stitching/pull/42).

**Comment by chinery** [comment]
Hi, yes this was fixed, apologies I forgot to update the original issue.
